local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,220 @@ fn test_slot_hash_expiry() {
+        if last_vote_on_a > common_ancestor_slot {
+            let blockstore = open_blockstore(&a_ledger_path);
+            info!(
+                "A majority fork: {:?}",
+                AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+            );
+            break;
+        }
+        sleep(Duration::from_millis(100));
+    }
+
+    // Kill A and restart B with voting. B should now fork off
+    info!("Killing A");
+    let a_info = cluster.exit_node(&a_pubkey);
+
+    info!("Restarting B");
+    b_info.config.voting_disabled = false;
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    // Allow B to fork now.
+    info!("Allowing B to fork");
+    loop {
+        sleep(Duration::from_millis(200));
+        let blockstore = open_blockstore(&b_ledger_path);
+        let (last_vote, _) = last_vote_in_tower(&b_ledger_path, &b_pubkey).unwrap();
+        let mut iterator = AncestorIterator::new(last_vote, &blockstore);
+        let first_ancestor = iterator.next().unwrap();
+        let second_ancestor = iterator.next().unwrap();
+        if second_ancestor > common_ancestor_slot
+            && first_ancestor % 4 == 3
+            && first_ancestor == second_ancestor + 1
+        {
+            break;
+        }
+    }
+
+    info!("Kill B");
+    b_info = cluster.exit_node(&b_pubkey);
+
+    info!("Resolve the partition");
+    {
+        // Here we let B know about the missing blocks that A had produced on its partition
+        let a_blockstore = open_blockstore(&a_ledger_path);
+        let b_blockstore = open_blockstore(&b_ledger_path);
+        copy_blocks(last_vote_on_a, &a_blockstore, &b_blockstore);
+    }
+
+    // Now restart A and B and see if B is able to eventually switch onto the majority fork
+    info!("Restarting A & B");
+    cluster.restart_node(&a_pubkey, a_info, SocketAddrSpace::Unspecified);
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    let blockstore = open_blockstore(&a_ledger_path);
+    info!(
+        "A majority fork: {:?}",
+        AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+    );
+    let last_vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+    let blockstore = open_blockstore(&b_ledger_path);
+    info!(
+        "B last vote: {} minority fork: {:?}",
+        last_vote_on_b,
+        AncestorIterator::new(last_vote_on_b, &blockstore).collect::<Vec<Slot>>()
+    );
+
+    for i in 0..100 {
+        sleep(Duration::from_millis(500));
+        let vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+        if vote_on_b > last_vote_on_b {
+            info!(
+                "B has voted again: {} last vote {} loop {}",
+                vote_on_b, last_vote_on_b, i,
+            );
+            break;
+        }
+    }
```

Comment:
If this doesn't happen within the 100 iterations, should probably panic?

Reply:
Done.